plugins/serialization/src/main/java/org/qbicc/plugin/serialization/BuildtimeHeap.java

```diff
@@ -84,18 +104,37 @@ public GlobalVariableElement getAndRegisterGlobalClassArray(ExecutableElement or
         return classArrayGlobal;
     }
 
-    public synchronized Data serializeStringLiteral(String value) {
-        // String literals are interned via equals, not ==
-        if (stringLiterals.containsKey(value)) {
-            return stringLiterals.get(value);
+    public synchronized Data serializeVmObject(VmObject value) {
+        if (vmObjects.containsKey(value)) {
+            return vmObjects.get(value);
         }
-        LoadedTypeDefinition jls = ctxt.getBootstrapClassContext().findDefinedType("java/lang/String").load();
-        Data sl = serializeObject(jls, value);
-        stringLiterals.put(value, sl);
+        PhysicalObjectType ot = value.getObjectType();
+        Data sl;
+        if (ot instanceof ClassObjectType) {
+            if (value instanceof VmClass) {
+                if (value instanceof VmArrayClass) {
+                    if (((VmArrayClass)value).getInstanceObjectType().getElementType() instanceof ReferenceArrayObjectType) {
```

Comment:
Trying to think about which code would trigger this case - is it something like:
```
Class MyClass {
    Class<?> arrayClass = MyClass[].class;
}
```

The restriction seems reasonable to start with but trying to get the implications clear in my head

Reply:
Yeah.  It's a staging thing.  We do support getting these class objects at runtime thanks to #639, but I didn't want to think about how to connect the build time creation of one of them to the runtime data structure yet.